### PR TITLE
Address training exploits

### PR DESCRIPTION
**Pull request by worm-girl**
#### Summary
Bugfixes "Limits dodge training from boomer bile and amigara horrors, reduces skill of sewer rats and crawlers"

#### Purpose of change

Several very old monsters had wild melee values that were either not what is laid out in game_balance.md, or they were not doing damage commensurate with their intended threat level. Player armor has gotten a lot more powerful in recent years.

Boomers, being unable to go through windows and possessing a non-damaging ranged attack, were training players up to 6+ dodge without ever damaging them. It was trivial to lead a boomer to an open window and then just stand there getting barfed on for six hours.

There's a similar old exploit where you can just corner a sewer rat and stand there forever while it tries to bite you and does no damage through even very basic armor, which will get you to 6 ranks. 

#### Describe the solution

- Adds an optional float that can be added any time on_dodge or dodge_check are called. The check will function as normal, but will treating the float as the hit roll/difficulty for training purposes.
- Boomer barf will train dodge up to 2. This is a bit better than you can do at home by throwing a tennis ball at a wall, because the boomer's movements are unpredictable. However, it's not a real threat in the situation described and training ought to be done elsewhere.
- Amigara horrors are described as two stories tall and are a boss monster at the finale of a dungeon. They previously had nine melee skill but did a laughable 2d4 damage, probably to balance out their paralyze ability.  This made it fairly trivial to armor up enough to be immune to their attacks, and then just chill out down there until you had 9 dodge. I think the horrors need a real buff, but in testing, they just get so many hits in that if you're not armored head to toe 100%, you're probably going to die. Rather than ruining these enemies for people engaging with the content normally, I dropped their melee skill down to 7 and added a check that prevents the character from attempting to blocking or dodging if they are paralyzed with fear. This does not substantially increase the monster's power level, as it was not being blocked and dodged by characters with the paralyze debuff, though they were trying and gaining experience.

A better fix would be to weaken the paralysis attack, give it some kind of counterplay as with #71460 and make the Horrors be slow monsters that could grab a helpless player. Out of scope for now.
- Reduced melee values for all non-mutated rats to 3. Rats are just basic little critters and wouldn't teach you how to fight hulks and monsters. If we want mean bitey rats, we should ramp up their damage.
- Crawler melee was reduced from 9 to 4. The high value never made any sense for a creature that seems intended to be so pathetic ( "A conglomeration of human parts fused together in a horrible mishmash of function, slowly dragging its eldritch body across the ground." ), and doesn't reflect what's listed in game_balance.md. If it has 9 melee it should be attacking very quickly or very forcefully, or both, not slowly dragging itself along the ground and doing 2d6 damage.
- Caps boomered and glowing (from huge boomer bile) effects at 30s and 6m respectively. While testing, I realized that it was possible to get them to stack ad infinitum if you're doing something like getting vominted on repeatedly.
 
#### Describe alternatives you've considered

Currently, dodging is basically a sure thing past a certain point. I think that should not be the case except with non-attack stuff like grabs. Or maybe we've just let dodge bonuses balloon out of control?

Rather than just checking for fear paralysis, we could give effects the ability to directly influence learn rate. If you're blackout drunk, in an adrenaline rush, or flying on experimental super-science stims, you may not remember a fight enough to learn as much from it, to say nothing of the non-combat skills.

Dark Wyrms also need an update, but messing with the Amigara Horrors showed that they need something more comprehensive than a damage boost. I was going to simply reduce their melee skill, but game_balance.md cites them as an example of something that ought to be very dangerous.

#### Testing

Spawned a survivor with leather body armor, a survivor hood, chitin arm guards, and a medium steel katana. With 8 stats and 3 in all skills, I narrowly won a fight with 3 amigara horrors where they all started at melee. They have a small amount of cut damage and while you're paralyzed are likely to hit exposed parts sooner or later. Things would have gone much smoother if I had better stats/skills, better gear, a gun, a friend, a dog, etc. It's a boss fight, so that feels fine to me.

Spawned a survivor, trapped a huge boomer in a house, had it vomit on me through the window. I got to 2 ranks, but no higher. Spawned a bloated corvid next to the boomer, saw the boomer barfed on the crow without anything unusual happening.

Also tested things with a regular boomer. Worked the same way.

#### Additional context

This doesn't fix everything. You can still throw a zombie in your basement and let it beat on you, but fixing that would likely require daily caps on skill gain, or capping skill gain per individual monster, or both IDK.

### CODE REVIEW

File: src/character.cpp [hunk 1]
```diff
@@ -1946,8 +1947,13 @@ void Character::on_dodge( Creature *source, float difficulty )
 
     // Even if we are not to train still call practice to prevent skill rust
     difficulty = std::max( difficulty, 0.0f );
-    practice( skill_dodge, difficulty * 2, difficulty );
 
+    // If training_level is set, treat that as the difficulty instead
+    if( training_level != 0.0 ) {
+        difficulty = training_level;
```
Comment (RenechCDDA) [inline src/character.cpp:1953]: It is not possible for this conditional to ever be fulfilled, nothing ever passes a float argument for training_level into these functions

File: src/creature.h [hunk 2]
```diff
@@ -497,9 +497,13 @@ class Creature : public viewer
         */
         void longpull( const std::string &name, const tripoint &p );
 
-        bool dodge_check( float hit_roll, bool force_try = false );
-        bool dodge_check( monster *z );
-        bool dodge_check( monster *z, bodypart_id bp, const damage_instance &dam_inst );
+        /**
+         * If training_level is anything but 0, the check will only train target's skill to that level
+        */
+        bool dodge_check( float hit_roll, bool force_try = false, float training_level = 0 );
```
Comment (RenechCDDA) [inline src/creature.h:503]:
```suggestion
        bool dodge_check( float hit_roll, bool force_try = false, float training_level = 0.0f );
        bool dodge_check( monster *z, float training_level = 0.0f );
        bool dodge_check( monster *z, bodypart_id bp, const damage_instance &dam_inst,
                          float training_level = 0.0f );
        
```

File: src/character.cpp [hunk 3]
```diff
@@ -2684,7 +2690,7 @@ bool Character::practice( const skill_id &id, int amount, int cap, bool suppress
 {
     SkillLevel &level = get_skill_level_object( id );
     const Skill &skill = id.obj();
-    if( !level.can_train() || in_sleep_state() ||
+    if( !level.can_train() || in_sleep_state() || has_effect( effect_fearparalyze ) ||
```
Comment (RenechCDDA) [inline src/character.cpp:2693]: This prevents them from being trained, but that is not a fix. Characters which are paralyzed should not be *trying to dodge at all*.

Comment (fairyarmadillo) [reply src/character.cpp:2693]: I don't know why I didn't think of that. I can put a check into melee.cpp for that stuff.

File: src/character.cpp (hunk 1 above)
Comment (fairyarmadillo) [reply src/character.cpp:1953]: I believe I've fixed this. It was working in testing but I definitely had 0 and not 0.0f in a few places.

**Pull request merged**